### PR TITLE
Update clang-tidy and clang-format to pip installed 16 versions

### PR DESCRIPTION
**Pull request by mlxd**
### Before submitting

Please complete the following checklist when submitting a PR:

- [ ] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      [`tests`](../tests) directory!

- [ ] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [ ] Ensure that the test suite passes, by running `make test`.

- [ ] Add a new entry to the `.github/CHANGELOG.md` file, summarizing the
      change, and including a link back to the PR.

- [ ] Ensure that code is properly formatted by running `make format`. 

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:**
The clang-tidy check keeps [failing the CI](https://github.com/PennyLaneAI/pennylane-lightning/actions/runs/6723825702/attempts/2?pr=547) as it read the wrong header files.

**Description of the Change:**
Bump clang-tidy and clang-format to v16 to resolve the issue. A few mods related to OpenMP macros had to be made to tie it together. 

**Benefits:**
Stable CI.

**Possible Drawbacks:**

**Related GitHub Issues:**

### CODE REVIEW

**Review by vincentmr — APPROVED** [review]
Looks good to me, thanks @mlxd .

File: bin/format [hunk 1]
```diff
@@ -18,7 +18,7 @@ CLANG_FMT_STYLE_CFG = {
 
 IGNORE_PATTERNS = ["external"]
 
-DEFAULT_CLANG_FORMAT_VERSION=14
+DEFAULT_CLANG_FORMAT_VERSION=16
```
Comment (AmintorDusko) [inline bin/format:21]: Should it be 17?

File: pennylane_lightning/core/src/simulators/lightning_qubit/gates/cpu_kernels/avx_common/AVX512Concept.hpp [hunk 2]
```diff
@@ -31,8 +31,12 @@ namespace Internal {
...
     static_assert(std::is_same_v<T, float> || std::is_same_v<T, double>);
 };
-template <> struct AVX512Intrinsic<float> { using Type = __m512; };
-template <> struct AVX512Intrinsic<double> { using Type = __m512d; };
+template <> struct AVX512Intrinsic<float> {
+    using Type = __m512;
+};
+template <> struct AVX512Intrinsic<double> {
+    using Type = __m512d;
+};
```
Comment (AmintorDusko) [inline pennylane_lightning/core/src/simulators/lightning_qubit/gates/cpu_kernels/avx_common/AVX512Concept.hpp:39]: This is so not compact. :neutral_face:

**Review by AmintorDusko — APPROVED** [review]
Thank you for that!
I see that versions are not 17, but 16. Should we update the PR title?
Can we have a brief description of all the proposed changes in the PR description.

File: bin/format (hunk 1 above)
Comment (vincentmr) [reply bin/format:21]: @mlxd tried 17 on at first, but decided on 16 in the end.

File: pennylane_lightning/core/src/simulators/lightning_qubit/gates/cpu_kernels/avx_common/TwoQubitGateHelper.hpp [hunk 3]
```diff
@@ -353,7 +353,9 @@ class TwoQubitGateWithoutParamHelper {
      */
     auto operator()(std::complex<Precision> *arr, const size_t num_qubits,
                     const std::vector<size_t> &wires, bool inverse) const
-        -> ReturnType requires SymmetricTwoQubitGateWithoutParam<AVXImpl> {
+        -> ReturnType
+        requires SymmetricTwoQubitGateWithoutParam<AVXImpl>
+    {
```
Comment (mlxd) [inline pennylane_lightning/core/src/simulators/lightning_qubit/gates/cpu_kernels/avx_common/TwoQubitGateHelper.hpp:358]: Hmmm, these shouldn't be on this line, but on the previous one.  Dame with everywhere below. If this isn't movable, I think we can accept for now and fix later, but I wonder why it isn't uniform across functions

Comment (vincentmr) [reply pennylane_lightning/core/src/simulators/lightning_qubit/gates/cpu_kernels/avx_common/TwoQubitGateHelper.hpp]: Should we just turn off clang-format for those lines?

**Pull request merged**